RFS/src/main/java/com/rfs/cms/IWorkCoordinator.java

```diff
@@ -0,0 +1,95 @@
+package com.rfs.cms;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+
+public interface IWorkCoordinator extends AutoCloseable {
+
+    /**
+     * Used as a discriminated union of different outputs that can be returned from acquiring a lease.
+     * Exceptions are too difficult/unsafe to deal with when going across lambdas and lambdas seem
+     * critical to glue different components together in a readable fashion.
+     */
+    interface WorkAcquisitionOutcome {
+        <T> T visit(WorkAcquisitionOutcomeVisitor<T> v) throws IOException;
+    }
+
+    interface WorkAcquisitionOutcomeVisitor<T> {
+        T onAlreadyCompleted() throws IOException;
+        T onNoAvailableWorkToBeDone() throws IOException;
+        T onAcquiredWork(WorkItemAndDuration workItem) throws IOException;
+    }
+
+    /**
+     * This represents that a work item was already completed.
+     */
+    class AlreadyCompleted implements WorkAcquisitionOutcome {
+        @Override public <T> T visit(WorkAcquisitionOutcomeVisitor<T> v) throws IOException {
+            return v.onAlreadyCompleted();
+        }
+    }
+
+    /**
+     *
+     */
+    class NoAvailableWorkToBeDone implements WorkAcquisitionOutcome {
+        @Override
+        public <T> T visit(WorkAcquisitionOutcomeVisitor<T> v) throws IOException {
+            return null;
```

Comment:
This looks like a bug, shouldn't this be `v.onNoAvailableWorkToBeDone`?

Reply:
👍